### PR TITLE
WIP PR for finetuning

### PR DESCRIPTION
**Pull request by abhigya-sodani**
Added finetuning for Neo LLM, chatgpt, and gpt.

Finetuning for Neo runs on the server well an then runs into threading issues after about 33% completion:

```
Exception in thread Thread-11:
Traceback (most recent call last):
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\llm_vm\completion\optimize.py", line 41, in new_thread
    t[0] = foo()
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\llm_vm\completion\optimize.py", line 278, in promise
 * Detected change in 'c:\\users\\abhigya sodani\\anaconda3\\Lib\\threading.py', reloading
    return succeed_train(use_completion)
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\llm_vm\completion\optimize.py", line 264, in actual_train
    asyncStart(models.MODELCONFIG.small_model.finetune(training_exs,self,c_id))
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\llm_vm\onsite_llm.py", line 224, in finetune
    trainer.train()
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\transformers\trainer.py", line 1645, in train
    return inner_training_loop(
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\transformers\trainer.py", line 2035, in _inner_training_loop
 * Detected change in 'c:\\users\\abhigya sodani\\anaconda3\\Lib\\site-packages\\transformers\\trainer.py', reloading
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\transformers\trainer.py", line 2321, in _maybe_log_save_evaluate
    metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\transformers\trainer.py", line 3053, in evaluate
    output = eval_loop(
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\transformers\trainer.py", line 3235, in evaluation_loop
    for step, inputs in enumerate(dataloader):
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 521, in __next__
    data = self._next_data()
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 561, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\torch\utils\data\_utils\fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\torch\utils\data\_utils\fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\llm_vm\onsite_llm.py", line 28, in __getitem__
Exception in thread Thread-2:
Traceback (most recent call last):
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 932, in _bootstrap_inner
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0000026DE9B87E30)

Thread 0x00008a78 (most recent call first):
<no Python frame>

Thread 0x00008724 (most recent call first):
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 306 in wait
  File "c:\users\abhigya sodani\anaconda3\lib\queue.py", line 179 in get
  File "c:\users\abhigya sodani\anaconda3\lib\site-packages\tensorboard\summary\writer\event_file_writer.py", line 227 in run
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 932 in _bootstrap_inner
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 890 in _bootstrap

Thread 0x00004200 (most recent call first):
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 890 in _bootstrap

Thread 0x00007fe8 (most recent call first):
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "c:\users\abhigya sodani\anaconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00002ec8 (most recent call first):
<no Python frame>
```

To better test finetuning, made client an actual programmatic interface, but then I would just get the completion and no information on the fine tuning tasks.

### CODE REVIEW

File: src/llm_vm/onsite_llm.py [hunk 1]
```diff
@@ -279,3 +359,30 @@ def generate(self,prompt, max_length=100,**kwargs): # both tokenizer and model t
             model="gpt-3.5-turbo-0301",
             **kwargs)
         return ans['choices'][0]['message']['content']
+
+    def finetune(self, dataset, optimizer, c_id):
+        old_model = optimizer.storage.get_model(c_id)
+        training_file = create_jsonl_file(dataset)
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:365]: Why are we writing the training data to disk?

File: src/llm_vm/onsite_llm.py [hunk 2]
```diff
@@ -173,6 +196,35 @@ def generate(self,prompt,max_length=100,**kwargs): # both tokenizer and model ta
...
+
+        trainer = Trainer(
+            model=self.model,
+            args=training_args,
+            train_dataset=FinetuningDataset(f_dataset,len(final_dataset)),
+            eval_dataset=FinetuningDataset(f_dataset,len(final_dataset)),
+            data_collator=data_collator,
+        )
+
+        trainer.train()
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:225]: Does this update the model in place ?

Comment (abhigya-sodani) [reply src/llm_vm/onsite_llm.py:225]: no creates a weights file that gets saved

File: src/llm_vm/onsite_llm.py (hunk 1 above)
Comment (abhigya-sodani) [reply src/llm_vm/onsite_llm.py:365]: Matt's code, creates a temp jsonl file that we send to openAI. I just moved this code out from optimize and made it a memeber of the gpt class under a finetune function

**Comment by ajn2004** [comment]
made redundant with pull #66

**Pull request closed without merging**